system_zero/src/alu/rotate_shift.rs

```diff
@@ -0,0 +1,596 @@
+}
+
+fn eval_rotate_shift<F: Field, P: PackedField<Scalar = F>>(
+    lv: &[P; NUM_COLUMNS],
+    yield_constr: &mut ConstraintConsumer<P>,
+    filter: P,
+) -> (P, P, P, P, P, P, P) {
+    // Input
+    let input_lo = lv[COL_ROTATE_SHIFT_INPUT_LO];
+    let input_hi = lv[COL_ROTATE_SHIFT_INPUT_HI];
+
+    // Delta is the shift/rotate displacement; exp is delta mod 32 or
+    // (32 - delta mod 32) mod 32, depending on whether the operation
+    // direction is left or right.
+    let exp_bits = COL_ROTATE_SHIFT_EXP_BITS.map(|r| lv[r]);
+    let top_bit = lv[COL_ROTATE_SHIFT_DELTA_DIV32];
+
+    let pow_exp_aux_0 = lv[COL_ROTATE_SHIFT_POW_EXP_AUX_0];
+    let pow_exp_aux_1 = lv[COL_ROTATE_SHIFT_POW_EXP_AUX_1];
+    let pow_exp_aux_2 = lv[COL_ROTATE_SHIFT_POW_EXP_AUX_2];
+    let pow_exp = lv[COL_ROTATE_SHIFT_POW_EXP];
+
+    let shifted_lo_0 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_0];
+    let shifted_lo_1 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_1];
+    let shifted_hi_0 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_0];
+    let shifted_hi_1 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_1];
+    let shifted_lo_aux_0 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_AUX_0];
+    let shifted_lo_aux_1 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_AUX_1];
+    let shifted_hi_aux_0 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_AUX_0];
+    let shifted_hi_aux_1 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_AUX_1];
+
+    // Output
+    let output_lo = lv[COL_ROTATE_SHIFT_OUTPUT_0];
+    let output_hi = lv[COL_ROTATE_SHIFT_OUTPUT_1];
+
+    // Check that every "bit" of exp_bits and top_bit is 0 or 1
+    exp_bits.map(|b| yield_constr.constraint(filter * (b * b - b)));
+    yield_constr.constraint(filter * (top_bit * top_bit - top_bit));
+
+    // Check that pow_exp = 2^exp, where exp is formed from the bits
+    // exp_bits[0..4].
+    //
+    // 2^exp = \prod_i=0^4 (2^(2^i) if exp_bits[i] = 1 else 1)
+    //       = \prod_i=0^4 ((2^(2^i) - 1) * exp_bits[i] + 1)
+    //       = pow_exp
+    //
+    // on the conditions that:
+    //
+    // pow_exp_aux_0 = \prod_i=0^1 ((2^i - 1) * exp_bits[i] + 1)
+    // pow_exp_aux_1 = \prod_i=2^3 ((2^i - 1) * exp_bits[i] + 1)
+    // pow_exp_aux_2 = pow_exp_aux_0 * pow_exp_aux_1
+    // pow_exp_mod32 = pow_exp_aux_2 * ((2^(2^4) - 1) * exp_bits[4] + 1)
+
+    let one = P::ONES;
+    // c[i-1] = 2^(2^i) - 1
+    let c = [1, 2, 3, 4].map(|i| P::from(F::from_canonical_u64(1u64 << (1u32 << i))) - P::ONES);
+
+    let constr1 = (exp_bits[0] + one) * (c[0] * exp_bits[1] + one);
+    yield_constr.constraint(filter * (constr1 - pow_exp_aux_0));
+    let constr2 = (c[1] * exp_bits[2] + one) * (c[2] * exp_bits[3] + one);
+    yield_constr.constraint(filter * (constr2 - pow_exp_aux_1));
+    let constr3 = pow_exp_aux_0 * pow_exp_aux_1;
+    yield_constr.constraint(filter * (constr3 - pow_exp_aux_2));
+    let constr4 = pow_exp_aux_2 * (c[3] * exp_bits[4] + one);
+    yield_constr.constraint(filter * (constr4 - pow_exp));
```

Comment:
I really suggest making this its own function. It’s kinda difficult to parse, and has a very different purpose from the rest of the function.

Reply:
Good call; found a significant bug in the recursive version when I refactored!